### PR TITLE
sql/sem/tree: break dependencies on catalog, sessiondata, security

### PR DESCRIPTION
**Pull request by ajwerner**
This pile of commits moves stuff and makes some minor changes to break more dependencies of tree. Most of the commits are totally mechanical. The ones which aren't relate to abstracting `SearchPath` a tad and moving cast options to a struct. This should be straightforward to review commit-by-commit.

The end goal here will be to break the dependencies of the parser on `util/log`. There's still a few more things remaining here and there, but this felt like a reasonable stopping point.

### CODE REVIEW

File: pkg/sql/sessiondata/search_path.go [hunk 1]
```diff
@@ -290,3 +291,19 @@ func (iter *SearchPathIter) Next() (path string, ok bool) {
...
 }
+
+// IterateSearchPath iterates the search path. If a non-nil error is
+// returned, iteration is stopped. If iterutils.StopIteration() is returned
+// from the iteration function,  a nil error is returned to the caller.
+func (sp *SearchPath) IterateSearchPath(f func(schema string) error) error {
+	iter := sp.Iter()
+	for schema, ok := iter.Next(); ok; schema, ok = iter.Next() {
+		if err := f(schema); err != nil {
+			if iterutil.Done(err) {
```
Comment (otan) [inline pkg/sql/sessiondata/search_path.go:302]: this iter package is cool

File: pkg/sql/alter_database.go [hunk 2]
```diff
@@ -35,6 +35,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlerrors"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqltelemetry"
+	"github.com/cockroachdb/cockroach/pkg/sql/username"
```
Comment (otan) [inline pkg/sql/alter_database.go]: this import confused me when i reviewed this commit but got lost somewhere. i don't know if you want to fix it.

File: pkg/sql/catalog/catconstants/schemas.go [hunk 3]
```diff
@@ -10,12 +10,15 @@
 
 package catconstants
 
-import "github.com/cockroachdb/cockroach/pkg/keys"
+// PublicSchemaID redefines keys.PublicSchemaID to avoid an import. It exists
+// to deal with time-travel queries from moments before the time when all
+// databases other than system were given a public schema.
+const PublicSchemaID = 29
```
Comment (otan) [inline pkg/sql/catalog/catconstants/schemas.go:16]: maybe it belongs to `pkg/base`??? :trollface:

**Review by otan — APPROVED** [review]

File: pkg/sql/alter_database.go (hunk 2 above)
Comment (ajwerner) [reply pkg/sql/alter_database.go]: I fixed this.

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Pull request merged**